# Movie Genre Classification

This project predicts the genre of a movie based on its plot summary using Machine Learning.

Dataset: IMDb Genre Classification Dataset from Kaggle.

Internship: CodSoft Machine Learning Internship

Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries imported successfully!")

Libraries imported successfully!


Load Training Data

In [2]:
train_data = pd.read_csv(
    "train_data.txt",
    sep=" ::: ",
    names=["ID", "TITLE", "GENRE", "DESCRIPTION"],
    engine="python"
)

train_data.head()

,ID,TITLE,GENRE,DESCRIPTION
0,1,Oscar et la dame rose (2009),drama,Listening in to a conversation between his doc...
1,2,Cupid (1997),thriller,A brother and sister with a past incestuous re...
2,3,"Young, Wild and Wonderful (1980)",adult,As the bus empties the students for their fiel...
3,4,The Secret Sin (1915),drama,To help their unemployed father make ends meet...
4,5,The Unrecovered (2007),drama,The film's title refers not only to the un-rec...


Check Dataset Shape

In [3]:
print("Rows and Columns:")
print(train_data.shape)

Rows and Columns:
(54214, 4)


Check Missing Values

In [4]:
train_data.isnull().sum()

,0
ID,0
TITLE,0
GENRE,0
DESCRIPTION,0


See Genre Distribution

In [5]:
train_data["GENRE"].value_counts().head(10)

,count
GENRE,
drama,13613
documentary,13096
comedy,7447
short,5073
horror,2204
thriller,1591
action,1315
western,1032
reality-tv,884


Prepare Features and Labels

In [6]:
X = train_data["DESCRIPTION"]
y = train_data["GENRE"]

print(X.head())
print("\n")
print(y.head())

0    Listening in to a conversation between his doc...
1    A brother and sister with a past incestuous re...
2    As the bus empties the students for their fiel...
3    To help their unemployed father make ends meet...
4    The film's title refers not only to the un-rec...
Name: DESCRIPTION, dtype: object


0       drama
1    thriller
2       adult
3       drama
4       drama
Name: GENRE, dtype: object


Convert Text into Numbers (TF-IDF)

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [8]:
tfidf = TfidfVectorizer(
    stop_words='english',
    max_features=5000
)

X_tfidf = tfidf.fit_transform(X)

print(X_tfidf.shape)

(54214, 5000)


Train-Test Split

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(
    X_tfidf,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(43371, 5000)
(10843, 5000)


Train Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression

In [12]:
lr_model = LogisticRegression(max_iter=1000)

lr_model.fit(X_train, y_train)

print("Model trained successfully!")

Model trained successfully!


Make Predictions

In [13]:
y_pred = lr_model.predict(X_test)

print(y_pred[:10])

['comedy' 'drama' 'documentary' 'horror' 'music' 'drama' 'documentary'
 'documentary' 'documentary' 'drama']


Calculate Accuracy

In [14]:
from sklearn.metrics import accuracy_score

In [15]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.5799133081250576


Classification Report

In [16]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      action       0.51      0.25      0.34       263
       adult       0.75      0.21      0.33       112
   adventure       0.42      0.14      0.21       139
   animation       0.60      0.09      0.15       104
   biography       0.00      0.00      0.00        61
      comedy       0.51      0.59      0.55      1443
       crime       0.29      0.02      0.04       107
 documentary       0.66      0.85      0.74      2659
       drama       0.54      0.78      0.64      2697
      family       0.39      0.07      0.12       150
     fantasy       0.00      0.00      0.00        74
   game-show       0.94      0.42      0.59        40
     history       0.00      0.00      0.00        45
      horror       0.64      0.56      0.60       431
       music       0.62      0.47      0.54       144
     musical       1.00      0.02      0.04        50
     mystery       0.00      0.00      0.00        56
        news       1.00    

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Test the Model

In [17]:
sample_movie = [
    "A group of friends enter a haunted house and are chased by evil spirits."
]

sample_vector = tfidf.transform(sample_movie)

prediction = lr_model.predict(sample_vector)

print("Predicted Genre:", prediction[0])

Predicted Genre: horror


In [18]:
sample_movie = [
    "A funny story about two friends who get into hilarious situations while traveling."
]

sample_vector = tfidf.transform(sample_movie)

prediction = lr_model.predict(sample_vector)

print("Predicted Genre:", prediction[0])

Predicted Genre: comedy


In [19]:
sample_movie = [
    "A retired soldier fights a dangerous criminal organization to save his family."
]

sample_vector = tfidf.transform(sample_movie)

prediction = lr_model.predict(sample_vector)

print("Predicted Genre:", prediction[0])

Predicted Genre: action


In [20]:
import joblib

In [21]:
joblib.dump(lr_model, "movie_genre_model.pkl")
joblib.dump(tfidf, "tfidf_vectorizer.pkl")

print("Model Saved Successfully!")

Model Saved Successfully!


Download Saved Files

In [22]:
from google.colab import files

files.download("movie_genre_model.pkl")
files.download("tfidf_vectorizer.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Train Naive Bayes

In [23]:
from sklearn.naive_bayes import MultinomialNB

nb_model = MultinomialNB()

nb_model.fit(X_train, y_train)

nb_pred = nb_model.predict(X_test)

nb_accuracy = accuracy_score(y_test, nb_pred)

print("Naive Bayes Accuracy:", nb_accuracy)

Naive Bayes Accuracy: 0.5225491100249009


Train SVM

In [24]:
from sklearn.svm import LinearSVC

svm_model = LinearSVC()

svm_model.fit(X_train, y_train)

svm_pred = svm_model.predict(X_test)

svm_accuracy = accuracy_score(y_test, svm_pred)

print("SVM Accuracy:", svm_accuracy)

SVM Accuracy: 0.5694918380521996


Compare All Models

In [25]:
print("Logistic Regression Accuracy:", accuracy)
print("Naive Bayes Accuracy:", nb_accuracy)
print("SVM Accuracy:", svm_accuracy)

Logistic Regression Accuracy: 0.5799133081250576
Naive Bayes Accuracy: 0.5225491100249009
SVM Accuracy: 0.5694918380521996


Choose Best Model

In [26]:
model_accuracies = {
    "Logistic Regression": accuracy,
    "Naive Bayes": nb_accuracy,
    "SVM": svm_accuracy
}

best_model_name = max(model_accuracies, key=model_accuracies.get)
best_accuracy = model_accuracies[best_model_name]

print("Best Model:", best_model_name)
print("Best Accuracy:", best_accuracy)

Best Model: Logistic Regression
Best Accuracy: 0.5799133081250576


Save the Best Model

In [27]:
import joblib

joblib.dump(lr_model, "movie_genre_model.pkl")
joblib.dump(tfidf, "tfidf_vectorizer.pkl")

print("Best model saved successfully!")

Best model saved successfully!


# Conclusion

Three machine learning models were trained and evaluated:

1. Logistic Regression
2. Naive Bayes
3. Support Vector Machine (SVM)

Results:

- Logistic Regression: 57.99%
- SVM: 56.95%
- Naive Bayes: 52.25%

Logistic Regression achieved the highest accuracy and was selected as the final model for movie genre prediction.